components/sup/src/manager/service/config.rs

```diff
@@ -938,4 +976,84 @@ mod test {
         cfg.default = Some(toml::from_str(default_toml).unwrap());
         assert_eq!(default_toml, toml::to_string(&cfg).unwrap());
     }
+
+    // env_key: the name of the environment variable the config should
+    //     be read from
+    // package_name: the name of the package that would read
+    //     environment configuration from `env_key`.
+    // input_config: the value of the environment variable `env_key`;
+    //     can be either JSON or TOML
+    // expected_config_as_toml: for validation purposes; this should
+    //     be the TOML version of `input_config`, since we always read to
+    //     TOML, regardless of the input format.
+    fn test_expected_successful_environment_parsing(
+        env_key: &str,
+        package_name: &str,
+        input_config: &str,
+        expected_config_as_toml: &str,
+    ) {
+        env::set_var(env_key, &input_config);
+
+        let expected = toml_from_str(expected_config_as_toml);
+        let result = Cfg::load_environment(&package_name.to_string());
+
+        // Clean up the environment after ourselves
+        env::remove_var(env_key);
+
+        match result {
+            Ok(Some(actual)) => {
+                assert_eq!(actual, expected);
+            }
+            _ => {
+                panic!("Expected '{:?}', but got {:?}", expected, result);
+            }
+        }
+    }
+
+    #[test]
+    fn can_parse_toml_environment_config() {
+        test_expected_successful_environment_parsing(
+            "HAB_TESTING_TOML",
+            "testing-toml",
+            "port = 1234",
+            "port = 1234",
+        );
+    }
+
+    #[test]
+    fn can_parse_json_environment_config() {
+        test_expected_successful_environment_parsing(
+            "HAB_TESTING_JSON",
+            "testing-json",
+            "{\"port\": 1234}",
+            "port = 1234",
+        );
+    }
+
+    #[test]
+    fn parse_environment_config_that_is_neither_toml_nor_json_fails() {
+        let key = "HAB_TESTING_TRASH";
+        let config = "{\"port: 1234 what even is this!!!!?! =";
+
+        env::set_var(key, &config);
+
+        let result = Cfg::load_environment(&"testing-trash".to_string());
+
+        // Clean up the environment after ourselves
+        env::remove_var(key);
+
+        assert!(result.is_err(), "Expected an error: got {:?}", result);
+    }
+
+    #[test]
+    fn no_environment_config_is_fine() {
+        match Cfg::load_environment(&"omg-there-wont-be-an-environment-variable-for-this".to_string())
```

Comment:
`export HAB_OMG_THERE_WONT_BE_AN_ENVIRONMENT_VARIABLE_FOR_THIS='in_your_face="haha"'`

Probably putting `=` in the string would defeat me. ;)